### DISCUSSION TITLE
NASA ESP-IDF: Compiler Errors

### DISCUSSION THREAD

**Discussion by DerTiger1** [omerfaruk-aran/esphome_samsung_hvac_bus#117] [General]
Compiling on esp-idf fails with current & also previous version (I was waiting for somebody else to report this ;-) )

`Compiling .pioenvs/esphome-klimaanlage/src/esphome/components/samsung_ac/protocol_non_nasa.o
In file included from src/esphome/components/samsung_ac/protocol_nasa.cpp:4:
src/esphome/components/samsung_ac/protocol_nasa.cpp: In function 'void esphome::samsung_ac::process_messageset(std::__cxx11::string, std::__cxx11::string, esphome::samsung_ac::MessageSet&, esphome::optional<std::set<short unsigned int> >&, esphome::samsung_ac::MessageTarget*)':
src/esphome/components/samsung_ac/protocol_nasa.cpp:584:31: error: format '%f' expects argument of type 'double', but argument 7 has type 'long int' [-Werror=format=]
                 ESP_LOGW(TAG, "s:%s d:%s VAR_in_temp_water_tank_f %f", source.c_str(), dest.c_str(), message.value);
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/core/log.h:72:36: note: in definition of macro 'ESPHOME_LOG_FORMAT'
 #define ESPHOME_LOG_FORMAT(format) format
                                    ^~~~~~
src/esphome/core/log.h:150:28: note: in expansion of macro 'esph_log_w'
 #define ESP_LOGW(tag, ...) esph_log_w(tag, __VA_ARGS__)
                            ^~~~~~~~~~
src/esphome/components/samsung_ac/protocol_nasa.cpp:584:17: note: in expansion of macro 'ESP_LOGW'
                 ESP_LOGW(TAG, "s:%s d:%s VAR_in_temp_water_tank_f %f", source.c_str(), dest.c_str(), message.value);
                 ^~~~~~~~
cc1plus: some warnings being treated as errors
*** [.pioenvs/esphome-klimaanlage/src/esphome/components/samsung_ac/protocol_nasa.o] Error 1`

**Comment by lanwin** [comment]
Thanks for reporting this. Are you on the stable or main branch. The main branch also has an IDF build witch did not fail.

**Comment by DerTiger1** [comment]
I'm on the main branch:
external_components:
  - source: github://lanwin/esphome_samsung_ac@main
  
 With recommended ESP-IDF:
 esp32:
  board: m5stack-atom
  framework:
    # type: arduino
    type: esp-idf
    version: recommended

**Reply by lanwin** [reply]
Strage. I added the github idf build especially to get such errors. But it did not catch this.

Can you recheck if it works now?

**Reply by DerTiger1** [reply]
Sorry - still doesn't compile, but with new failure:

Compiling .pioenvs/esphome-klimaanlage/src/esphome/components/samsung_ac/samsung_ac_device.o
In file included from src/esphome/components/samsung_ac/samsung_ac_device.cpp:1:
src/esphome/components/samsung_ac/samsung_ac_device.cpp: In member function 'void esphome::samsung_ac::Samsung_AC_Climate::set_alt_mode_by_name(esphome::samsung_ac::ProtocolRequest&, const AltModeName&)':

**These lines are marked red - it seems to be invalid variable declaration/assignment const instead of char:**

**_src/esphome/components/samsung_ac/samsung_ac_device.cpp:151:23: error: format '%s' expects argument of type 'char*', but argument 5 has type 'const AltModeName' {aka 'const std::__cxx11::basic_string<char>'} [-Werror=format=]_**

========================= [FAILED] Took 272.89 seconds =========================

debugging:

[https://github.com/esphome/esphome/blob/dev/esphome/core/log.h](url)
`151: #define ESP_LOGW(tag, ...) esph_log_w(tag, __VA_ARGS__)

`115: #if ESPHOME_LOG_LEVEL >= ESPHOME_LOG_LEVEL_WARN
#define esph_log_w(tag, format, ...) \
  esp_log_printf_(ESPHOME_LOG_LEVEL_WARN, tag, __LINE__, ESPHOME_LOG_FORMAT(format), ##__VA_ARGS__)

`56: void esp_log_printf_(int level, const char *tag, int line, const char *format, ...)  // NOLINT
    __attribute__((format(printf, 4, 5)));
#ifdef USE_STORE_LOG_STR_IN_FLASH
void esp_log_printf_(int level, const char *tag, int line, const __FlashStringHelper *format, ...);
#endif
void esp_log_vprintf_(int level, const char *tag, int line, const char *format, va_list args);  // NOLINT
#ifdef USE_STORE_LOG_STR_IN_FLASH
void esp_log_vprintf_(int level, const char *tag, int line, const __FlashStringHelper *format, va_list args);
#endif
#if defined(USE_ESP32_FRAMEWORK_ARDUINO) || defined(USE_ESP_IDF)
int esp_idf_log_vprintf_(const char *format, va_list args);  // NOLINT
#endif

### ==> That's why %s expects to be char ;-)`

EDIT: I'm interested to know where the variable name is declared ??? (I'm studying the code with my knowings of C++ 35years ago :-) )

ESP_LOGW(TAG, "Unsupported alt_mode %s", name);
`

**Reply by DerTiger1** [reply]
I've fixed it:
with "%d" it compile fine ;-)

btw: did you remove the "Outdoor temperature"?
     # outdoor_temperature: # Should be used with outdoor device address
      #  name: "Klima_Aussentemperatur"

**Reply by lanwin** [reply]
Please try again. %d is not correct. This line uses an std::string which did not work here. It must be an s_str. 

Nop Outdoor temperature should still be there. Where do you miss it?

**Reply by DerTiger1** [reply]
I can confirm that it compiles now and outdoor temperature is working for me again.
THX
